src/libsystemd-network/sd-dhcp6-client.c

```diff
@@ -727,23 +727,32 @@ static int client_parse_message(
                 DHCP6Message *message,
                 size_t len,
                 sd_dhcp6_lease *lease) {
+        size_t pos = 0;
         int r;
-        uint8_t *optval, *option, *id = NULL;
-        uint16_t optcode, status;
-        size_t optlen, id_len;
         bool clientid = false;
-        be32_t iaid_lease;
+        uint8_t *id = NULL;
+        size_t id_len;
 
         assert(client);
         assert(message);
         assert(len >= sizeof(DHCP6Message));
         assert(lease);
 
-        option = (uint8_t *)message + sizeof(DHCP6Message);
         len -= sizeof(DHCP6Message);
 
-        while ((r = dhcp6_option_parse(&option, &len, &optcode, &optlen,
-                                       &optval)) >= 0) {
+        while (pos < len) {
+                DHCP6Option *option = (DHCP6Option *)&message->options[pos];
+                uint16_t optcode, optlen, status;
+                uint8_t *optval;
+                be32_t iaid_lease;
+
+                if (len < sizeof(DHCP6Option) || len < sizeof(DHCP6Option) + be16toh(option->len))
```

Comment:
nitpick: in theory i think `offsetof(DHCP6Option, data) + …` is nicer than `sizeof(DHCP6Option) + …` because the former removes alignment artifacts. In this specific case there are no alignment whitepsaces hence it's fine, but out of a matter of principle I'd probably always prefer offsetof() over sizeof() in cases like this